src/System.Private.CoreLib/shared/System/MemoryExtensions.Trim.cs

```diff
@@ -0,0 +1,817 @@
+            where T : IEquatable<T>
+        {
+            // Initially, start==len==0. If ClampStart trims all, start==len
+            Debug.Assert((uint)start <= span.Length);
+
+            int end = span.Length - 1;
+
+            if (trimElement != null)
+            {
+                for (; end >= start; end--)
+                {
+                    if (!trimElement.Equals(span[end]))
+                    {
+                        break;
+                    }
+                }
+            }
+            else
+            {
+                for (; end >= start; end--)
+                {
+                    if (span[end] != null)
+                    {
+                        break;
+                    }
+                }
+            }
+
+            return end - start + 1;
+        }
+
+        /// <summary>
+        /// Removes all leading and trailing occurrences of a set of elements specified
+        /// in a readonly span from the memory.
+        /// </summary>
+        /// <param name="memory">The source memory from which the elements are removed.</param>
+        /// <param name="trimElements">The span which contains the set of elements to remove.</param>
+        /// <remarks>If <paramref name="trimElements"/> is empty, the memory is returned unaltered.</remarks>
+        public static Memory<T> Trim<T>(this Memory<T> memory, ReadOnlySpan<T> trimElements)
+            where T : IEquatable<T>
+        {
+            // Use nested checks to avoid unnecessary branching for the unlikely case of N <= 1
+            if (trimElements.Length <= 1)
+            {
+                return trimElements.Length == 0 ? memory : Trim(memory, trimElements[0]);
+            }
+
+            ReadOnlySpan<T> span = memory.Span;
+            int start = ClampStart(span, trimElements);
+            int length = ClampEnd(span, start, trimElements);
+            return memory.Slice(start, length);
+        }
+
+        /// <summary>
+        /// Removes all leading occurrences of a set of elements specified
+        /// in a readonly span from the memory.
+        /// </summary>
+        /// <param name="memory">The source memory from which the elements are removed.</param>
+        /// <param name="trimElements">The span which contains the set of elements to remove.</param>
+        /// <remarks>If <paramref name="trimElements"/> is empty, the memory is returned unaltered.</remarks>
+        public static Memory<T> TrimStart<T>(this Memory<T> memory, ReadOnlySpan<T> trimElements)
+            where T : IEquatable<T>
+        {
+            // Use nested checks to avoid unnecessary branching for the unlikely case of N <= 1
+            if (trimElements.Length <= 1)
+            {
+                return trimElements.Length == 0 ? memory : TrimStart(memory, trimElements[0]);
+            }
+
+            return memory.Slice(ClampStart(memory.Span, trimElements));
```

Comment:
Nit: this could be written with less nesting as:
```C#
if (trimElements.Length > 1)
{
    return memory.Slice(ClampStart(memory.Span, trimElements));
}

if (trimElements.Length == 1)
{
    return TrimStart(memory, trimElements[0]);
}

return memory;
```
or even just:
```C#
return
    trimElements.Length > 1 ? memory.Slice(ClampStart(memory.Span, trimElements) :
    trimElements.Length == 1 ? TrimStart(memory, trimElements[0]) :
    memory;
```

Reply:
That's cleaner, I will fix. Would a switch add any benefit here - 0, 1, default?

Reply:
Bench: `switch` is ever so slightly faster, but not statistically significant (ratio = 0.99 consistently). 
I will see which one looks cleaner.